### ISSUE TITLE
[QUERY] How to mock raising events from ServiceBusSessionProcessor in a unit test

### ISSUE THREAD

**Issue by AnnejanBarelds** [Azure/azure-sdk-for-net#35660]
### Library name and version

Azure.Messaging.ServiceBus 7.13.1

### Query/Question

Hi,

_TL;DR: It looks like ServiceBusSessionProcessor cannot be mocked such that event handling just works, even though event handlers are the way to interact with it from consumer code. Is that by design? If so, how would you write a unit test that validates the correctness of the event handlers that are **actually registered to ServiceBusSessionProcessor**?_

I'm having difficulty mocking a ServiceBusSessionProcessor in a unit test. That is, I can mock its methods, but I can't raise events or intercept handler registration, as events are not virtual. Most observations below also apply to ServiceBusProcessor.

I hope I'm missing something trivial, and that there is a simple way to validate what I would want my test to validate. I'll try to work out the problem and my own attempts thus far at resolving them.

The code to test is a IHostedService implementation that gets a ServiceBusClient injected, which it uses to create a ServiceBusSessionProcessor. It wires up handlers to the ProcessMessageAsync and ProcessErrorAsync events, and starts the processor. Upon receiving of a message, it extracts the message payload from the body, and passes that to a an external handler, which is also injected. The event handlers are private members.

The code looks like this:

    public class MessageProcessor : IHostedService
    {
        private readonly MessageProcessorOptions _options;
        private readonly ServiceBusClient _serviceBusClient;
        private readonly IMessageHandler _messageHandler;
        private ServiceBusSessionProcessor? _processor;

        public MessageProcessor(IOptions<MessageProcessorOptions> options, ServiceBusClient serviceBusClient, IMessageHandler messageHandler)
        {
            _options = options.Value;
            _serviceBusClient = serviceBusClient;
            _messageHandler = messageHandler;
        }

        public async Task StartAsync(CancellationToken cancellationToken)
        {
            if ((_processor?.IsProcessing).GetValueOrDefault()) return;

            var processorOptions = new ServiceBusSessionProcessorOptions
            {
                AutoCompleteMessages = false
            };
            _processor = _serviceBusClient.CreateSessionProcessor(_options.Queue, processorOptions);
            _processor.ProcessMessageAsync += ProcessMessageAsync;
            _processor.ProcessErrorAsync += ProcessErrorAsync;
            await _processor.StartProcessingAsync(cancellationToken);
        }

        public async Task StopAsync(CancellationToken cancellationToken)
        {
            if (_processor != null && !_processor.IsClosed)
            {
                await _processor.StopProcessingAsync(cancellationToken);
                await _processor.CloseAsync(cancellationToken);
            }
        }

        private Task ProcessErrorAsync(ProcessErrorEventArgs arg)
        {
            // Log stuff
            return Task.CompletedTask;
        }

        private async Task ProcessMessageAsync(ProcessSessionMessageEventArgs arg)
        {
            var payload = JsonSerializer.Deserialize<MessagePayload>(arg.Message.Body);
            await _messageHandler.Handle(payload);
            await arg.CompleteMessageAsync(arg.Message);
        }
    }

I made several attempts to make this work.

### Naive mock
The first attempt is a simple mock of all the dependencies:

        [Fact]
        public async Task GivenMockProcessor_WhenMessageReceived_ShouldCallMessageHandler()
        {
            // Arrange
            var message = ServiceBusModelFactory.ServiceBusReceivedMessage(BinaryData.FromObjectAsJson(new MessagePayload()));
            var mockProcessor = new Mock<ServiceBusSessionProcessor>();
            var mockClient = new Mock<ServiceBusClient>();
            mockClient
                .Setup(client => client.CreateSessionProcessor(It.IsAny<string>(), It.IsAny<ServiceBusSessionProcessorOptions>()))
                .Returns(mockProcessor.Object);
            Mock<ServiceBusSessionReceiver> mockSessionReceiver = new();

            TimerCallback dispatchMessage = async _ =>
            {
                ProcessSessionMessageEventArgs processArgs = new(
                    message: message,
                    receiver: mockSessionReceiver.Object,
                    cancellationToken: CancellationToken.None);

                mockProcessor.Raise(processor => processor.ProcessMessageAsync += (args) => { return Task.CompletedTask; }, processArgs);
            };

            Timer messageDispatchTimer = new(
                dispatchMessage,
                null,
                Timeout.Infinite,
                Timeout.Infinite);

            mockProcessor.Setup(processor => processor.StartProcessingAsync(It.IsAny<CancellationToken>())).Returns(Task.CompletedTask).Callback(() => messageDispatchTimer.Change(0, 1000));
            mockProcessor.Setup(processor => processor.StopProcessingAsync(It.IsAny<CancellationToken>())).Returns(Task.CompletedTask).Callback(() => messageDispatchTimer.Change(Timeout.Infinite, Timeout.Infinite));

            var handlerMock = new Mock<IMessageHandler>();
            handlerMock.Setup(h => h.Handle(It.IsAny<MessagePayload>())).Verifiable();

            var options = Options.Create(new MessageProcessorOptions { Queue = "theQueue" });

            var sut = new MessageProcessor(options, mockClient.Object, handlerMock.Object);

            // Act
            await sut.StartAsync(CancellationToken.None); // This throws a null pointer when an attempt is made to register an event handler, as is done in the SUT
            await Task.Delay(2000);
            await sut.StopAsync(CancellationToken.None);

            // Assert
            handlerMock.Verify(h => h.Handle(It.IsAny<MessagePayload>()), Times.AtLeastOnce());
        }
This breaks in the IHostedService implementation, because the registration of a handler of ProcessMessageAsync is greeted with a NullReferenceException. So that's the first question:

**_Is registering an event handler on a mocked ServiceBusSessionProcessor supposed to yield a NullReferenceException?_**

### Mocking the InnerProcessor
In an attempt to fix this, I went into the source code and found the protected InnerProcessor property. So I tried to mock that as well:

        [Fact]
        public async Task GivenMockProcessorWithInnerProcessor_WhenMessageReceived_ShouldCallMessageHandler()
        {
            // Arrange
            var message = ServiceBusModelFactory.ServiceBusReceivedMessage(BinaryData.FromObjectAsJson(new MessagePayload()));
            var mockProcessor = new Mock<ServiceBusSessionProcessor>();
            var mockClient = new Mock<ServiceBusClient>();
            mockClient
                .Setup(client => client.CreateSessionProcessor(It.IsAny<string>(), It.IsAny<ServiceBusSessionProcessorOptions>()))
                .Returns(mockProcessor.Object);
            Mock<ServiceBusSessionReceiver> mockSessionReceiver = new();

            TimerCallback dispatchMessage = async _ =>
            {
                ProcessSessionMessageEventArgs processArgs = new(
                    message: message,
                    receiver: mockSessionReceiver.Object,
                    cancellationToken: CancellationToken.None);

                mockProcessor.Raise(processor => processor.ProcessMessageAsync += (args) => { return Task.CompletedTask; }, processArgs); // Throws (kills the test host process); complains about non-virtual members
            };

            Timer messageDispatchTimer = new(
                dispatchMessage,
                null,
                Timeout.Infinite,
                Timeout.Infinite);

            mockProcessor.Setup(processor => processor.StartProcessingAsync(It.IsAny<CancellationToken>())).Returns(Task.CompletedTask).Callback(() => messageDispatchTimer.Change(0, 1000));
            mockProcessor.Setup(processor => processor.StopProcessingAsync(It.IsAny<CancellationToken>())).Returns(Task.CompletedTask).Callback(() => messageDispatchTimer.Change(Timeout.Infinite, Timeout.Infinite));

            var handlerMock = new Mock<IMessageHandler>();
            handlerMock.Setup(h => h.Handle(It.IsAny<MessagePayload>())).Verifiable();

            var options = Options.Create(new MessageProcessorOptions { Queue = "theQueue" });

            var sut = new MessageProcessor(options, mockClient.Object, handlerMock.Object);

            // Act
            await sut.StartAsync(CancellationToken.None);
            await Task.Delay(2000);
            await sut.StopAsync(CancellationToken.None);

            // Assert
            handlerMock.Verify(h => h.Handle(It.IsAny<MessagePayload>()), Times.AtLeastOnce());
        }
This breaks upon raising the event using Moq Raise().

### Using Moq to setup the event handler
An alternate approach to at least get to the registered handle and then invoke that manually, instead of invoking it through raising the event from the ServiceBusSessionProcessor also broke down:

        [Fact]
        public async Task GivenMockProcessorWithEventSetup_WhenMessageReceived_ShouldCallMessageHandler()
        {
            // Arrange
            var message = ServiceBusModelFactory.ServiceBusReceivedMessage(BinaryData.FromObjectAsJson(new MessagePayload()));
            Func<ProcessSessionMessageEventArgs, Task> handler = null;

            var mockProcessor = new Mock<ServiceBusSessionProcessor>();
            mockProcessor
                .SetupAdd(processor => processor.ProcessMessageAsync += (args) => { return Task.CompletedTask; }) // Throws; complains about non-virtual members
                .Callback<Func<ProcessSessionMessageEventArgs, Task>>(h => handler = h);
            var mockClient = new Mock<ServiceBusClient>();
            mockClient
                .Setup(client => client.CreateSessionProcessor(It.IsAny<string>(), It.IsAny<ServiceBusSessionProcessorOptions>()))
                .Returns(mockProcessor.Object);
            Mock<ServiceBusSessionReceiver> mockSessionReceiver = new();

            TimerCallback dispatchMessage = async _ =>
            {
                ProcessSessionMessageEventArgs processArgs = new(
                    message: message,
                    receiver: mockSessionReceiver.Object,
                    cancellationToken: CancellationToken.None);

                mockProcessor.Raise(processor => processor.ProcessMessageAsync += (args) => { return Task.CompletedTask; }, processArgs); // Throws (kills the test host process); complains about non-virtual members
            };

            Timer messageDispatchTimer = new(
                dispatchMessage,
                null,
                Timeout.Infinite,
                Timeout.Infinite);

            mockProcessor.Setup(processor => processor.StartProcessingAsync(It.IsAny<CancellationToken>())).Returns(Task.CompletedTask).Callback(() => messageDispatchTimer.Change(0, 1000));
            mockProcessor.Setup(processor => processor.StopProcessingAsync(It.IsAny<CancellationToken>())).Returns(Task.CompletedTask).Callback(() => messageDispatchTimer.Change(Timeout.Infinite, Timeout.Infinite));

            var handlerMock = new Mock<IMessageHandler>();
            handlerMock.Setup(h => h.Handle(It.IsAny<MessagePayload>())).Verifiable();

            var options = Options.Create(new MessageProcessorOptions { Queue = "theQueue" });

            var sut = new MessageProcessor(options, mockClient.Object, handlerMock.Object);

            // Act
            await sut.StartAsync(CancellationToken.None);
            await Task.Delay(2000);
            await sut.StopAsync(CancellationToken.None);

            // Assert
            handlerMock.Verify(h => h.Handle(It.IsAny<MessagePayload>()), Times.AtLeastOnce());
        }

This breaks upon setting up the events on the mock using Moq SetupAdd().

So both Moq approaches (Raise() and SetupAdd()) for doing interesting stuff with events broke down. As it turns out, the ProcessMessageAsync and ProcessErrorAsync events are not virtual(!). From what I understood, virtual public members are the way to go in all Azure SDK's for reasons of testability, since SDK types no longer have interfaces for reasons of evolvability. So, the second question is:

**_Why are events on the ServiceBusProcessor and ServiceBusSessionProcessor not marked as virtual? What alternative are you offering for testability?_**

I've seen the sample code at https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/tests/Samples/Sample15_MockingClientTypes.cs (more specifically the MockRunningTheSessionProcessor() test), but that sample assumes outside access to the event handlers and it invokes them manually. I have two objections to this approach:
1. I don't like my event handler methods to be publicly accessible. The code under test is consuming a dependency from which it obtains a message which it passes to another downstream dependency. From the perspective of the test code, I shouldn't have to know or care about how that works internally.
2. I don't like the idea of invoking the handlers manually, even apart from their visibility. Invoking the handlers manually means invoking the handlers that I _assume_ to be attached to the events, _not_ the handlers I _know_ to be attached. A refactoring where the handler method is replaced and now contains a bug, will now be missed. I'm testing the old handler, whereas at runtime another faulty handler is actually wired to the event.

So the third question is:

**_Is there anything that can be done to mock ServiceBusSessionProcessor (or ServiceBusProcessor for that matter) such that a unit test can be written to validate the correctness of the event handlers that will actually be called at runtime?_**

### Environment

Using .NET 6.0; Visual Studio 2022 17.5.3; Moq 4.18.4; xUnit 2.4.2

**Comment by jsquire** [comment]
Hi @AnnejanBarelds.   Thank you for reaching out and we regret that you're experiencing difficulties.   Generally, we recommend calling your handlers directly to test your logic.    The Service Bus client library guarantees that these will be called at the appropriate times.  

Since you've mentioned that you'd prefer not to take that route, I'd suggest referring to the [Simulating the processor running](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample15_MockingClientTypes.md#simulating-the-processor-running) section of our [Mocking client types sample](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample15_MockingClientTypes.md).   Further context and discussion is available there.

**Comment by AnnejanBarelds** [comment]
Hi @jsquire and @JoshLove-msft. Thanks for the replies.

As for the content referenced by @jsquire: I've seen that, as I mentioned in my question. The problem with that approach is that it does not guarantee that the handlers I'm invoking are also the handlers that have been registered. So it doesn't provide a way to test the correctness of how things are wired together.

With regards to the virtual OnProcessSessionMessage method mentioned by @JoshLove-msft: I've been playing with that too, in the last couple of days. Don't remember why I couldn't get it to work, but I'll give it another try and come back with my findings.

And more in general: however the mocking of events will turn out, there's still the NullReferenceException that is thrown by my first crack at a test. That's not related to raising an event or invoking a handler, but merely to registering the handler in the first place. Surely that should be possible with a mock without having to also mock a protected member? So at least for this issue, this is not yet resolved.

**Comment by JoshLove-msft** [comment]
> With regards to the virtual OnProcessSessionMessage method mentioned by @JoshLove-msft: I've been playing with that too, in the last couple of days. Don't remember why I couldn't get it to work, but I'll give it another try and come back with my findings.

This [test](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/tests/Processor/ProcessorTests.cs#L321) may be useful as a reference.

**Comment by jsquire** [comment]
Apologies; that's my error.  I thought that example invoked the event via the processor and had forgotten that we were doing so via the handler directly.    To use that with the registered handler for the processor, you'd need to the protected methods that Josh had mentioned.   

The best way to do so would be to create a custom processor type for testing that extends `ServiceBusProcessor` and exposes public methods to invoke the `OnXXX` ones.   If you're not able to extend the processor for testing, you'd need to do a bit of reflection to get at them.

Assuming you have a processor type named `processor`, something like:

```csharp
// Capture the protected methods that raise the event.
var processMessage = processor
    .GetType()
    .GetMethod("OnProcessMessageAsync", BindingFlags.Instance | BindingFlags.NonPublic)
     .CreateDelegate<Func<ProcessMessageEventArgs, Task>>(processor);

var processError = processor
     .GetType()
     .GetMethod("OnProcessErrorAsync", BindingFlags.Instance | BindingFlags.NonPublic)
     .CreateDelegate<Func<ProcessErrorEventArgs, Task>>(processor);

// Invoke the handler for processing a message.
var message = ServiceBusModelFactory.ServiceBusReceivedMessage(
    body: new BinaryData("message"),
    messageId: "messageId",
    correlationId: "correlationId");
        
var args = new ProcessMessageEventArgs(message, Mock.Of<ServiceBusReceiver>(), "Test", default);
await processMessage(args);
```

**Comment by AnnejanBarelds** [comment]
Thanks, that works!

Not sure why my earlier fiddling with these OnXXX methods failed; probably has something to do with all the different permutations of mock setups I was trying out. One of those permutations was probably my failure to set the InnerProcessor on a session processor, which you are also doing in the [MockSessionProcessor](https://github.com/Azure/azure-sdk-for-net/blob/2b592b4da3f6cab40e8e548dc7259bb71d3010a9/sdk/servicebus/Azure.Messaging.ServiceBus/tests/Processor/SessionProcessorTests.cs#L410). The consequence of this particular requirement is that using Moq to mock a ServiceBusSessionProcessor is not suitable in most real-life scenarios. I.e. it works in this [sample](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample15_MockingClientTypes.md#simulating-the-processor-running), but only because no actual event handlers are registered; mocking away a dependency for a real-life piece of code that actually registers event handlers will break with a NullReferenceException. Note that this is only true for ServiceBusSessionProcessors.

So all in all, this mostly looks like a documentation issue. Nudging people in the direction of using e.g. Moq to create processor mocks, especially in the case of session processors, is probably not too helpful. Instead, it would be great if the approach you guys shared in this thread becomes part of the documentation. It solves the NullReferenceException, and allows for testing the registered handler.

Documentation would be especially helpful because it's not the most intuitive approach. People tend to use frameworks such as Moq for creating mocks, and since that's also advertised in the current docs, crafting 'manual' mocks quickly feels like "I'm probably doing something wrong here".

Anyway, thanks again for pointing me in the right direction!

**Comment by jsquire** [comment]
Glad to hear that you're unblocked, @AnnejanBarelds, and thank you for your feedback.   Yours is a fairly unique scenario; this is the first time that we've seen a case where an application wants to validate that it has wired up the correct handlers.  While we certainly want to ensure that this is possible, it requires a non-trivial amount of complexity that isn't needed for the majority case.   

In mainstream scenarios for the processor, testing/mocking should be focused on the application-provided code in the event handlers and confirming that the processor is started/stopped as intended.  This is where our samples are focused - trying to guide developers towards testing their application code directly rather than indirectly by asserting control over the processor infrastructure.   I agree that it would be helpful to have something that discusses creating a mock-by-extension, but I'm not sure that including it in our mainstream sample is the right place for it.    We'll discuss this internally and work to come up with an approach.   I've opened #35696 to track.

I'm going to mark this as addressed.  Please feel free to unresolve if you'd like to continue the discussion.

**Comment by AnnejanBarelds** [comment]
Great, I'll read along.

Maybe it's indeed a minority scenario - even though I haven't met that many developers who think it's a good idea to make members public just for the sake of being able to unit test the class. But that is what the current guidance is asking for. And yes, marking them internal and visible to the test project is an improvement, but in general, dev's prefer to test the public surface area of a unit only.

Lastly, my sample tests above were simplifications. In reality, this is part of a component test (or integration test) where I want to have the entire app in memory and only mock out the outer edges, as validation that the entire dependency tree, including wiring up the handlers, works as intended. Much like a web app developer would use a [WebApplicationFactory](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1?view=aspnetcore-7.0) for end-to-end tests.
